### ISSUE TITLE
How to do relative imports in Python simply while allowing for elpy-shell-send-region-or-buffer (C-c C-c)

### ISSUE THREAD

**Issue by bgoodr** [jorgenschaefer/elpy#1109]
Thanks for this wonderful Emacs package!

I'm trying to:

 1. Load modules that are always to be located _relative_ to the non-module Python script that uses them, _and_,
 1. Be able to use the `C-c C-c` Elpy binding for development, _and_
 1. Not set `PYTHONPATH` in the environment. I might wish to copy the non-module file, _and_ its module subdirectory, to a different directory, and experiment with changes to that module subdirectory, _without_ having Python try to use a different module from the `PYTHONPATH` (e.g., confusion would result).

I have something working but it kludgy, and I would like to know if
anyone has come up with something simpler and thus more maintainable.

TL;DR: Here are the details:

I have the following non-module file stored at `~/mybin/ccc.py`:

    import sys
    import os
    my_dir = None
    try:
        my_dir = os.path.dirname(os.path.abspath(__file__))
    except:
        # For when using C-c C-c from Emacs elpy package, to avoid messiness with setting PYTHONPATH:
        my_dir = os.getcwd()
    sys.path.insert(0, os.path.join(my_dir, "my_python_modules"))
    # Annoyance: The next line causes flake8 to emit: E402 module level import not at top of file
    # Hence my desire to not have the bulky try/except clause above!:
    from basic_util import die
    
    die("death")

The contents of `~/mybin/my_python_modules/basic_util.py` is:

    import sys
    def die(errmsg):
        print "ERROR: {}".format(errmsg)
        sys.exit(1)

This works, but it is cumbersome since I have to copy and paste that
`try/except` boilerplate code across multiple non-module script files
residing inside `~/mybin/`. I would rather like to compress the above
boilerplate down to one or two lines of Python code, something of the
form:

    from <MAGIC_STUFF>my_python_modules import die

I've tried using a `<MAGIC_STUFF>` of ".". In `~/mybin/ddd.py` I have tried:

    import sys
    import os
    from .my_python_modules.basic_util import die
    die("death")

but that gives me:

    $ (cd /; python ~/mybin/ddd.py)
    Traceback (most recent call last):
      File "/home/brentg/mybin/ddd.py", line 18, in <module>
        from .my_python_modules.basic_util import die
    ValueError: Attempted relative import in non-package
    
I can set `PYTHONPATH` inside the environment, but that means I have
to hardcode paths to the `my_python_modules` directory, or I have to
use some Rube Goldberg-esc silly contraption involving
`elpy-mode-hook` functions which is more complicated than just living
with the copy/paste-ing of the boilerplate code.

Is there an easier way to go about this?

References:

 1. http://stackoverflow.com/questions/67631/how-to-import-a-module-given-the-full-path
 1. http://stackoverflow.com/questions/72852/how-to-do-relative-imports-in-python/73149#73149

**Comment by jorgenschaefer** [comment]
Would using `add2virtualenv ~/mybin/my_python_modules/` (from virtualenvwrapper) solve this for you?

**Comment by bgoodr** [comment]
Maybe, but what I didn't mention, and should have, is that some subset of my non-module .py files need to execute in whatever currently installed Python is available, and not require the user to run it within a virtualenv.  So I would like to find a solution that works both with, and without, virtualenv's.

If there is simply no better way to go about it than using virtualenv's or using my concocted kludgy boilerplate as described, I will just have to accept it and we can close this item.  

Also, after reading through the [Python 2.7 import machinery](https://docs.python.org/2.7/reference/simple_stmts.html#import), this is probably just a general Python issue and not specific to `elpy`. If so, then I apologize for belaboring it with you. :)

**Comment by jorgenschaefer** [comment]
Python finds modules by looking in `sys.path`. You want modules to be found, you have to add them to that path. You can do that using virtualenvs, `PYTHONPATH`, or by russnning the python interpreter in the given directory (because `""` is in there). You reject the first and second solution, so you could try the third. You can either manually start a Python shell using `M-x run-python` in the right directory, or use `os.chdir` to set the directory, I guess …

**Comment by bgoodr** [comment]
Understood. Running the python interpreter by changing the current working
directory is also problematic, since the current working directory for the
user is expected to not to silently change during execution for most/all
scripts: Imagine the user specifying an option of `-infile ./that_file`
only to get a failure where `./that_file` does not exist in the directory
that was changed to internally.

So I recommend to close this Issue as will not fix. I'll give it some more
thought.

Thanks a lot for your help!!
-Brent

**Comment by jorgenschaefer** [comment]
Ok – good luck with finding a good solution :-)

**Comment by dangom** [comment]
@bgoodr Any chances you came up with a better workflow in the meantime?

I'm facing the same issue and working around it by installing everything I develop (python setup.py develop) and ditching relative imports completely. Don't know if that's the best practice though.

**Comment by bgoodr** [comment]
@dangom  No. Unfortunately, I'm still using the same kludge as indicated.

**Comment by bgoodr** [comment]
After migrating to Python3, I ran into relative imports problem again, but this time using the "." syntax in the "from...import" statement

Since I see a lot of Python code nowadays using that relative import syntax, and feeling that it is the correct way to express code succinctly, I gave up and posted a question to SE: 

https://stackoverflow.com/questions/55895012/python-3-script-using-relative-imports-on-standard-input-gives-error-no-module

Hopefully we'll get some answers. Otherwise, I'm going to have to resort to explicitly setting PYTHONPATH to every directory elpy drives me to and doing away with the "." syntax (which is what I believe @dangom is concluding also.  That's sad that I would have to do that, but I'm at my wits end with this, after several hours poured into trying various workarounds.

**Comment by galaunay** [comment]
Would using `import` instead of sending the code work for you ?

This function would do that from Emacs:
```Elisp
(defun elpy-send-buffer-with-relative-imports ()
  (interactive)
  (elpy-shell--ensure-shell-running)
  (save-buffer)
  (python-shell-send-string
   (format "import os\nos.chdir('''%s''')\nfrom %s.%s import *;print('')"
           (elpy-project-root)
           (file-name-nondirectory (directory-file-name (elpy-project-root)))
           (file-name-base (buffer-file-name)))))
```

(It will only work for packages with one level, but it can be improved)

**Comment by bgoodr** [comment]
I'm not sure, since (elpy-project-root) is nil for me.

**Comment by bgoodr** [comment]
But I get your point. Just doing this from the python interactive buffer for that script does give some results:

    >>> import os
    >>> os.getcwd()
    '/tmp/topdir/subdir'
    >>> os.chdir('..')
    >>> os.getcwd()
    '/tmp/topdir'
    >>> from subdir.subscript1 import *
    subscript2 loaded successfully
    subscript1 loaded successfully
    >>> 

(of course, I'd have to use os.chdir() again to restore back to that current working directory.

**Comment by bgoodr** [comment]
I found a variation of the above idea and updated my question on SE : https://stackoverflow.com/questions/55895012/python-3-script-using-relative-imports-on-standard-input-gives-error-no-module

Search for "selfdir" to see that variation. It involves a rather kludgy workaround whereby a "selfdir" link is added to the directory to point to ../subdir where "subdir" is the tail directory of the current working directory.  That's a kludge as it leaves a symbolic link in the current working directory. So that really isn't elegant solution, but might work. 

The problem with that approach that I can see right away is that it implies that I'll be invoking a python interpreter each time, which is not what C-c C-c is currently doing (it binds to `elpy-shell-send-region-or-buffer`).

**Comment by galaunay** [comment]
> I'm not sure, since (elpy-project-root) is nil for me.

Elpy will detect your package as a project if it can find a `setup.py`, `setup.cfg` or `pyproject.toml` file at the package root.
Does it work if you add an empty `pyproject.toml` file at the package root ?

**Comment by bgoodr** [comment]
> Elpy will detect your package as a project if it can find a `setup.py`, `setup.cfg` or `pyproject.toml` file at the package root.
> Does it work if you add an empty `pyproject.toml` file at the package root ?

If "it" in "Does it work" mean `elpy-send-buffer-with-relative-imports` defined above, I've not gotten that far. When I open up the python file that is in the same directory as an empty `pyproject.toml`, then `(elpy-project-root)` still returns nil.

But I might not even need to depend upon `(elpy-project-root)` should mean if it is non-nil. I know what the parent directory is and could chdir to it, then chdir back from it.  But that might open up some problems: I do a lot of python development with elpy whereby I'm just running temporary scripts that load files in that very same directory that the .py file is in. If I were to start using a script that loads files from the parent directory, the logic will break since then the files are not in the expected locations (due to the chdir).

**Comment by galaunay** [comment]
> If "it" in "Does it work" mean `elpy-send-buffer-with-relative-imports` defined above.

Yes it was, sorry if I was unclear.

> When I open up the python file that is in the same directory as an empty pyproject.toml, then (elpy-project-root) still returns nil.

This is a unexpected. Elpy should internally use `elpy-project-find-python-root` to check for the presence of a `pyproject.toml` file (assuming you didn't customized `elpy-project-root-finder-functions` ?).  

> But I might not even need to depend upon (elpy-project-root) even if if it were to return non-nil. I know what the parent directory is and could chdir to it, then chdir back from it.

I was trying to find a more generic solution, that could be implemented in Elpy to allow users to evaluate scripts with relative imports. In your specific case, I guess you can just remove `elpy-project-root` and hard-code the path you want to change to.

> I don't want to have to continuously think about what type of import statements are being used (e.g., relative imports or absolute imports). 

Checking for relative import in the current script should be relatively easy. So you could write a function that will choose between `elpy-send-buffer` and `elpy-send-buffer-with-relative-imports` automatically depending on that.

**Comment by bgoodr** [comment]
I spent a lot of time looking at clever alternatives to allow me to do what I want, and all of the solutions I could think of were just too complicated to put into practice, so:

I ended up avoiding this by using symbolic links in the same directory in which the top-level script resides.  It's not great, but its better than having to change PYTHONPATH in the working environment.  So if I have utilitiy modules stored in my ~/bin/python directory, but I'm actually working on a script in, say, ~/some_project_dir/my_top_level.py, I create a symbolic link like this:

    ln -s ~/bin/python ~/some_project_dir/python

then inside the my_top_level.py file, I use import statements such as:

    from python.file_utils import read_lines_from_file

    lines = read_lines_from_file("/tmp/some_file")

So if I ever relocate ~/some_project_dir to some other directory, the symbolic link goes with it. If I ever were to move the ~/bin/python directory, then I do have the chore of updating the symbolic links, but that is a minor chore, and doesn't happen all that much.
